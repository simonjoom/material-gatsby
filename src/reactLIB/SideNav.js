import React, { Component } from 'react';
import cx from 'classnames';
import PropTypes from 'prop-types'; 
import idgen from './idgen';

class SideNav extends Component {
  constructor(props) {
    super(props);
    var r = idgen();
    this.id = props.id || `sidenav_${r}`;
  }

  componentDidMount() { 
    const { options } = this.props;
     Zepto(document).ready(function() {
        Zepto('.sidenav').sidenav(options);
      });
  }


  render() {
    const { className, children, trigger, fixed, ...props } = this.props;
    delete props.id;
    delete props.options;
    const classNames = cx(
      'sidenav',
      'sidenav-close',
      { 'sidenav-fixed': fixed || !trigger },
      className
    );

    return (
      <div>
        {this.renderTrigger()}
        <ul id={this.id} className={classNames} {...props}>
          {children}
        </ul>
      </div>
    );
  }

  renderTrigger() {
    const { trigger, fixed } = this.props;
    if (!trigger) {
      return;
    }
    const triggerView = fixed ? 'hide-on-large-only' : 'show-on-large';
    const classNames = cx(
      trigger.props.className,
      triggerView,
      'sidenav-trigger'
    );
    return React.cloneElement(trigger, {
      ref: t => (this._trigger = `[data-target=${this.id}]`),
      'data-target': this.id,
      className: classNames
    });
  }
}

SideNav.propTypes = {
  /**
   * Adds sidenav-fixed class to sidenav
   */
  fixed: PropTypes.bool,
  /**
   * sidenav id. If none is passed, an id will be generated.
   */
  id: PropTypes.string,
  /**
   * Component that is rendered to trigger the sidenav
   */
  trigger: PropTypes.node,
  /**
   * Options hash for the sidenav.
   * More info: http://materializecss.com/side-nav.html#options
   */
  options: PropTypes.shape({
    edge: PropTypes.oneOf(['left', 'right']),
    draggable: PropTypes.bool
  }),
  /**
   * Additional classes added to 'sidenav'
   */
  className: PropTypes.string,
  children: PropTypes.node
};

export default SideNav;
