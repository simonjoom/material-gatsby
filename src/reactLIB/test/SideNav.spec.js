import React from 'react';
import { shallow, mount } from 'enzyme';
import SideNav from '../src/SideNav';
import mocker from './helper/mocker';

function setup(props = {}, children, mounted) {
  const propsIn = {
    trigger: <button className="trigger">Show sidenav</button>,
    ...props
  };
  const component = <SideNav {...propsIn}>{children}</SideNav>;
  const wrapper = mounted ? mount(component) : shallow(component);
  const sideNav = wrapper.find('.sidenav');
  const trigger = wrapper.find('.sidenav-trigger');
  const sideNavProps = sideNav.props();
  const triggerProps = trigger.props();

  return {
    propsIn,
    wrapper,
    sideNav,
    trigger,
    sideNavProps,
    triggerProps
  };
}

describe('<SideNav />', () => {
  const sideNavMock = jest.fn();
  const restore = mocker('sidenav', sideNavMock);

  afterAll(() => {
    restore();
  });

  test('renders', () => {
    const { sideNav } = setup();
    expect(sideNav).toMatchSnapshot();
  });

  test('can be `fixed`', () => {
    const { sideNav } = setup({ className: 'red', fixed: true });
    expect(sideNav).toMatchSnapshot();
  });

  test('should render fixed if not passed trigger or fixed', () => {
    const { wrapper } = setup({ className: 'red' });
    expect(wrapper).toMatchSnapshot();
  });

  test('should render trigger on large screens not fixed`', () => {
    const { trigger } = setup({ className: 'green' });
    expect(trigger).toMatchSnapshot();
  });

  test('should be fixed if no trigger is passed', () => {
    const component = <SideNav className="red" />;
    const wrapper = shallow(component);
    const sideNav = wrapper.find('.sidenav');
    expect(sideNav).toMatchSnapshot();
  });

  test('should render a given id', () => {
    const { sideNavProps, triggerProps } = setup({ id: 'test' });
    expect(sideNavProps.id).toEqual('test');
    expect(triggerProps['data-target']).toEqual('test');
  });

  test('should render children', () => {
    const { sideNav } = setup({}, <span className="test-child" />);
    const children = sideNav.find('.test-child');

    expect(children).toHaveLength(1);
  });

  test('should consume props it uses', () => {
    const { sideNavProps } = setup({
      id: 'test123',
      options: {},
      shouldTransfer: true
    });

    expect(sideNavProps.trigger).toBeUndefined();
    expect(sideNavProps.options).toBeUndefined();
    expect(sideNavProps.shouldTransfer).toBeTruthy();
  });

  test('should call sideNav with the given options', () => {
    const options = {
      closeOnClick: true,
      edge: 'right'
    };
    mount(<SideNav trigger={<span>trigger</span>} options={options} />);
    expect(sideNavMock).toHaveBeenCalledWith(options);
  });
});
