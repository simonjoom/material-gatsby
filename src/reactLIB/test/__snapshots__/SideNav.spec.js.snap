// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SideNav /> can be \`fixed\` 1`] = `
<ul
  className="sidenav sidenav-fixed red"
  id="sidenav_1"
/>
`;

exports[`<SideNav /> renders 1`] = `
<ul
  className="sidenav"
  id="sidenav_0"
/>
`;

exports[`<SideNav /> should be fixed if no trigger is passed 1`] = `
<ul
  className="sidenav sidenav-fixed red"
  id="sidenav_4"
/>
`;

exports[`<SideNav /> should render fixed if not passed trigger or fixed 1`] = `
<div>
  <button
    className="trigger show-on-large sidenav-trigger"
    data-target="sidenav_2"
  >
    Show sidenav
  </button>
  <ul
    className="sidenav red"
    id="sidenav_2"
  />
</div>
`;

exports[`<SideNav /> should render trigger on large screens not fixed\` 1`] = `
<button
  className="trigger show-on-large sidenav-trigger"
  data-target="sidenav_3"
>
  Show sidenav
</button>
`;
